### PR TITLE
Handle an empty query parameter with @Default

### LINKED ISSUE #5515: Handle an empty query parameters with `@Default`
Say we have an annotated service that uses `@Default` annotation to optionally bind query parameters.
```java
class MyService {
    @Get("/list")
    public String list(@Default @Param Integer page) {
        ...
    }
}
```

As per the Javadoc of `@Default`, when a value is not specified for `@Default`, a non-existent parameter is converted into `null`. `page` is null if `/list` is the request path.
https://github.com/line/armeria/blob/ccb29fee0fea5a8657fe3ad2de1fe0a19062818c/core/src/main/java/com/linecorp/armeria/server/annotation/Default.java#L35-L36

However, it does not work well if there is a parameter but no value. A `NumberFormatException` is raised when `/list?page=` is sent as the request path.
```java
java.lang.NumberFormatException: For input string: ""
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Integer.parseInt(Integer.java:671)
	at java.base/java.lang.Integer.valueOf(Integer.java:988)
	at com.linecorp.armeria.internal.server.annotation.AnnotatedServiceTypeUtil.stringToType(AnnotatedServiceTypeUtil.java:177)
	at com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.convert(AnnotatedValueResolver.java:1021)
```

It would be better to solve this problem as follows the value of the parameter is empty.
- Inject an empty string ("") if the type of `page` is `String`. 
- Inject a null value if the type of `page` is not `String` such as `Integer`, `Boolean` and so on.

### PR DESCRIPTION
**Pull request by facewise**
Motivation:

Say we have an annotated service that uses `@Default `annotation to optionally bind query parameters.

```java
class MyService {
    @Get("/list")
    public String list(@Default @Param Integer page) {
        ...
    }
}
```

It does not inject `null` if there is a parameter but no value. A NumberFormatException is raised when `/list?page=` is sent as the request path.

It would be better to solve this problem as follows the value of the parameter is empty.

Modifications:

- Logger will warn that the parameter value is not present (It may be intended, such as sending an empty string as the parameter)
- Inject an empty string `""` if the type of the parameter is `String`
- Inject `null` if the type of the parameter is not `String` such as `Integer`, `Boolean` and so on
- Throw an `IllegalArgumentException` if the type of the parameter is primitive type
- Inject an empty list `[]` if the type of the parameter is `List` or `Set`

Result:

- Closes #5515.

### CODE REVIEW

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java [hunk 1]
```diff
@@ -86,7 +86,11 @@ class AnnotatedValueResolverTest {
     static final Set<String> existingHttpParameters = ImmutableSet.of("param1",
                                                                       "enum1",
                                                                       "sensitive");
-
+    // These parameters will be present without the values. e.g., ?emptyParam1=&emptyParam2=&...
+    static final Set<String> existingWithoutValueParameters = ImmutableSet.of("emptyParam1",
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java:90]: Should we also add an e2e test in which a client sends a query parameter with a value or an empty value? `EmptyParameterWithDefaultAnnotationTest` could be added for that.

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java [hunk 2]
```diff
@@ -1026,6 +1031,17 @@ private Object convert(@Nullable String value) {
         if (value == null) {
             return defaultOrException();
         }
+        if (value.isEmpty()) {
+            logger.warn("Parameter is present but the value is missing: " + httpElementName);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:1035]: As it is an expected behavior, we can remove the warning.

**Review by ikhoon — COMMENTED** [review]
The overall changes look nice.

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java (hunk 2 above)
Comment (facewise) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:1035]: I added this warning because @ikhoon commented in #5515 :
>In addition to the bugfix, we also need to improve the error message when an incorrect request is made so that users can easily identify the problem.

How about warning only if the type is not `String`?

```suggestion
            if (String.class.isAssignableFrom(elementType)) {
                return value;
            }
            if (elementType.isPrimitive()) {
                throw new IllegalArgumentException("Cannot set null to primitive type parameter: " +
                                                   httpElementName);
            }
            logger.warn("Parameter is present but the value is missing: " + httpElementName);
            return null;
```

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java (hunk 1 above)
Comment (facewise) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java:90]: Sure. Why not.

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/EmptyParameterWithDefaultAnnotationTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        }
+    }
+
+    private static final String NULL = "null";
+
+    private final WebClient client = WebClient.of(server.httpUri() + "/default");
+
+    @Test
+    void testEmptyStringParameterIsEmpty() {
+        AggregatedHttpResponse res = aggregate(client.get("/string?param=v"));
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/internal/server/annotation/EmptyParameterWithDefaultAnnotationTest.java:83]: Instead, we can use `server.blockingClient()`.

**Review by ikhoon — APPROVED** [review]
Nice work! 🚀 I can't believe that this is your first contribution. 😄

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/EmptyParameterWithDefaultAnnotationTest.java (hunk 3 above)
Comment (facewise) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/EmptyParameterWithDefaultAnnotationTest.java:83]: @ikhoon
I failed to replace it into `server.blockingCleint()` because it seems that the `DefaultBlockingWebClient` has `/` as `absolutePathRef()`.

This fails with 404 Not Found:
```java
final BlockingWebClient blockingWebClient = server.blockingWebClient();
AggregatedHttpResponse res = blockingWebClient.get("/string", 
                                                   QueryParams.of("param", "v"));
```

Should I append all the prefixes in front of request url like:
```java
final BlockingWebClient blockingWebClient = server.blockingWebClient();
AggregatedHttpResponse res = blockingWebClient.get("/default/string", 
                                                   QueryParams.of("param", "v"));
```
?

Comment (jrhee17) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/EmptyParameterWithDefaultAnnotationTest.java:83]: We don't have an easy way to specify a prefix for a prefix path in conjunction with `ServerExtension` at the moment - I think as-is is fine

File: core/src/main/java/com/linecorp/armeria/server/annotation/Default.java [hunk 4]
```diff
@@ -35,7 +37,19 @@
      * When {@link Default} annotation exists but {@link Default#value()} is not specified, {@code null}
      * value would be set if the parameter is not present in the request.
      *
-     * {@link Default} annotation is not allowed for a path variable. If a user uses {@link Default}
+     * <p>When {@link Default} annotation exists and {@link Default#value()} is not specified, but parameter is
+     * present in the request without the value e.g., {@code /foo?bar=}, then actions below will be processed
+     * depending on the type of the parameter.
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/server/annotation/Default.java:42]:
```suggestion
     * <p>When {@link Default} annotation exists and {@link Default#value()} is not specified, the parameter
     * may be present in the request without a value (e.g. {@code /foo?bar=}). In this case, the behavior
     * depends on the java type with which the parameter will be bound to.
```

**Review by minwoox — APPROVED** [review]
Great! Thanks!
